nitorch/tools/affine_reg.py

```diff
@@ -0,0 +1,961 @@
+    """Check device and set some Torch stuff.
+    """
+    opt['device'] = torch.device(opt['device'] if torch.cuda.is_available() else 'cpu')
+    torch.set_grad_enabled(False)
+    torch.backends.cudnn.benchmark = True
+
+    return opt
+
+
+def _optimise(q, Nq, args, opt):
+    """Run optimisation routine to fit q.
+
+    Parameters
+    ----------
+    q : (N, Nq) tensor_like
+        Lie algebra of affine registration fit.
+    Nq : int
+        Number of Lie basis.
+    args : tuple
+        All arguments for optimiser (except the parameters to be fitted)
+    opt : dict
+        See run_affine_reg.
+
+    Returns
+    ----------
+    q : (N, Nq) tensor_like
+        Optimised Lie algebra of affine registration fit.
+
+    """
+    if opt['optimiser'] == 'powell':
+        # Powell optimisation
+        callback = None
+        if opt['mean_space']:
+            # TODO: zero mean
+            callback = lambda x: _zero_mean(x, Nq)
+        s = q.shape
+        q = fmin_powell(_compute_cost, q, args=args, disp=False, callback=callback)
+        q = q.reshape(s)
+
+    return q
+
+
+def _show_results(q, args, fig_ax, opt):
+    """ Simple function for visualising some results during algorithm execution.
+    """
+    if opt['verbose']:
+        c, res = _compute_cost(q, *args, return_res=True)
+        print('_compute_cost({})={}'.format(opt['cost_fun'], c))
+        fig_ax = show_slices(res, fig_ax=fig_ax, fig_num=1, cmap='coolwarm')
+        # nii = nib.Nifti1Image(res.cpu().numpy(), M_fix.cpu().numpy())
+        # dir_out = os.path.split(pths[0])[0]
+        # nib.save(nii, os.path.join(dir_out, 'res_coreg.nii'))
+
+    return fig_ax
+
+
+def _smooth_for_reg(dat, vx, opt):
+    """Smoothing for image registration. FWHM is computed from voxel size
+       and sub-sampling amount.
+
+    Parameters
+    ----------
+    dat : (X, Y, Z) tensor_like
+        3D image volume.
+    vx : (3,) tensor_like
+        Voxel size.
+    opt : dict
+        See run_affine_reg.
+
+    Returns
+    -------
+    dat : (Nx, Ny, Nz) tensor_like
+        Smoothed 3D image volume.
+
+    """
+    # Get final sub-sampling level
+    samp = opt['samp']
+    if not isinstance(samp, tuple):
+        samp = (samp,)
+    samp = samp[-1]
+    if samp < 1:
+        samp = 1
+    samp = torch.tensor((samp,) * 3, dtype=dat.dtype, device=dat.device)
+    # Make smoothing kernel
+    fwhm = torch.sqrt(torch.max(samp ** 2 - vx ** 2, torch.zeros(3, device=dat.device, dtype=dat.dtype))) / vx
+    smo = smooth(('gauss',) * 3, fwhm=fwhm, device=dat.device, dtype=dat.dtype, sep=True)
+    # Padding amount for subsequent convolution
+    size_pad = (smo[0].shape[2], smo[1].shape[3], smo[2].shape[4])
+    size_pad = (torch.tensor(size_pad) - 1) // 2
+    size_pad = tuple(size_pad.int().cpu().tolist())
+    # Smooth deformation with Gaussian kernel (by separable convolution)
+    dat = pad(dat, size_pad, side='both')
+    dat = dat[None, None, ...]
+    dat = F.conv1d(dat, smo[0])
```

Comment:
Did you try that? For some reason I thought that `conv1d` needed (batch, channel, width) inputs. I would have used `conv3d` even though the kernel really is 1d (and I think that `smooth` actually returns 3d kernels, with shape (channel_in, channel_out, 1, 1, width), (channel_in, channel_out, 1, height, 1) etc)

Reply:
Hmm..good catch. I am not sure if it worked before, I think so, because I looked at the input which looked smoothed..

I changed to using conv3d anyway, and I also changed to conv2d in _hist_2d, for smoothing the joint histogram. It used to do conv1d too..